proxy/src/auth/backend/console.rs

```diff
@@ -37,61 +55,89 @@ pub enum GetAuthInfoError {
     BadSecret,
 
     #[error(transparent)]
-    Transport(TransportError),
+    ApiError(ApiError),
+}
+
+// This allows more useful interactions than `#[from]`.
+impl<E: Into<ApiError>> From<E> for GetAuthInfoError {
+    fn from(e: E) -> Self {
+        Self::ApiError(e.into())
+    }
 }
 
 impl UserFacingError for GetAuthInfoError {
     fn to_string_client(&self) -> String {
         use GetAuthInfoError::*;
         match self {
+            // We absolutely should not leak any secrets!
+            // Besides, user is unlikely to care about this detail.
             BadSecret => REQUEST_FAILED.to_owned(),
-            Transport(e) => e.to_string_client(),
+            // However, API might return a meaningful error.
+            ApiError(e) => e.to_string_client(),
         }
     }
 }
 
-impl<E: Into<TransportError>> From<E> for GetAuthInfoError {
-    fn from(e: E) -> Self {
-        Self::Transport(e.into())
-    }
-}
-
 #[derive(Debug, Error)]
 pub enum WakeComputeError {
-    // We shouldn't show users the address even if it's broken.
     #[error("Console responded with a malformed compute address: {0}")]
-    BadComputeAddress(String),
+    BadComputeAddress(Box<str>),
 
     #[error(transparent)]
-    Transport(TransportError),
+    ApiError(ApiError),
+}
+
+// This allows more useful interactions than `#[from]`.
+impl<E: Into<ApiError>> From<E> for WakeComputeError {
+    fn from(e: E) -> Self {
+        Self::ApiError(e.into())
+    }
 }
 
 impl UserFacingError for WakeComputeError {
     fn to_string_client(&self) -> String {
         use WakeComputeError::*;
         match self {
+            // We shouldn't show user the address even if it's broken.
+            // Besides, user is unlikely to care about this detail.
             BadComputeAddress(_) => REQUEST_FAILED.to_owned(),
-            Transport(e) => e.to_string_client(),
+            // However, API might return a meaningful error.
+            ApiError(e) => e.to_string_client(),
         }
     }
 }
 
-impl<E: Into<TransportError>> From<E> for WakeComputeError {
-    fn from(e: E) -> Self {
-        Self::Transport(e.into())
+/// This is essentially a [`Result`] but with a different
+/// [`Deserialize`] impl that makes more sense in our case.
+#[derive(Deserialize, Debug)]
+// Note that order of variants in untagged enums is important.
+#[serde(untagged)]
+enum ConsoleResponse<T> {
+    /// Nested payload.
+    Ok(T),
+    /// Console API error as text.
+    Err { error: Box<str> },
```

Comment:
@ololobus I suppose it's fine to show this text to user, but I'd like to hear a confirmation.

Reply:
@tychoish wdyt?

Reply:
> I suppose it's fine to show this text to user, but I'd like to hear a confirmation.

I **thought** so too, but I've just had a look on the management API server code. In the public v2 we have a middleware that only renders messages for the known semantic error, but I don't see something similar in the management API (I was looking on `ChiErrorHandler`). Thus, I assume it can throw away any internal error text (it also responds with `StatusBadRequest`, so I don't know how to distinguish errors), cc @antonyc correct me if I'm wrong

Reply:
In that case the only thing I could show user is `Internal error`, because I can't differentiate between "user doesn't exist", "not enough monies" or "warning: very secret error message" etc.

Reply:
> it can throw away any internal error text (it also responds with StatusBadRequest, so I don't know how to distinguish errors)
You are correct. Management API can expose any error to the user. We can add support for semantic errors with almost zero effort (just copy-paste some code).

Reply:
I think we need it, so that proxy was able to distinguish what it should present to the user (likely all non-500s) and what to hide (500s)

Reply:
Discussed this with @antonyc in private. I'm going to rework some parts.